File: SudokuSolver_Java_No_Streams/src/main/sudokusolver/javanostreams/logic/simple/PruneCandidates.java
package sudokusolver.javanostreams.logic.simple;

import sudokusolver.javanostreams.Board;
import sudokusolver.javanostreams.Cell;
import sudokusolver.javanostreams.RemoveCandidates;
import sudokusolver.javanostreams.SolvedCell;
import sudokusolver.javanostreams.SudokuNumber;
import sudokusolver.javanostreams.UnsolvedCell;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;

/*
 * If a cell is solved, then no other cells in the same unit can have that number as a candidate.
 */
public class PruneCandidates {
    public static List<RemoveCandidates> pruneCandidates(Board<Cell> board) {
        var removals = new ArrayList<RemoveCandidates>();
        for (var cell : board.getCells()) {
            if (cell instanceof UnsolvedCell unsolved) {
                var toRemove = EnumSet.noneOf(SudokuNumber.class);
                collectCandidates(board.getRow(cell.row()), toRemove);
                collectCandidates(board.getColumn(cell.column()), toRemove);
                collectCandidates(board.getBlock(cell.block()), toRemove);
                toRemove.retainAll(unsolved.candidates());
                if (!toRemove.isEmpty()) {
                    removals.add(new RemoveCandidates(unsolved, toRemove));
                }
            }
        }
        return removals;
    }

    private static void collectCandidates(List<Cell> unit, EnumSet<SudokuNumber> toRemove) {
        for (var cell : unit) {
            if (cell instanceof SolvedCell(_, _, var value)) {
                toRemove.add(value);
            }
        }
    }
}

File: SudokuSolver_Kotlin/src/main/sudokusolver/kotlin/logic/simple/PruneCandidates.kt
package sudokusolver.kotlin.logic.simple

import sudokusolver.kotlin.Board
import sudokusolver.kotlin.Cell
import sudokusolver.kotlin.RemoveCandidates
import sudokusolver.kotlin.SolvedCell
import sudokusolver.kotlin.SudokuNumber
import sudokusolver.kotlin.UnsolvedCell
import sudokusolver.kotlin.enumIntersect
import java.util.EnumSet

/*
 * If a cell is solved, then no other cells in the same unit can have that number as a candidate.
 */
fun pruneCandidates(board: Board<Cell>): List<RemoveCandidates> =
    board.cells.filterIsInstance<UnsolvedCell>().mapNotNull { cell ->
        val sameRow = board.getRow(cell.row)
        val sameColumn = board.getColumn(cell.column)
        val sameBlock = board.getBlock(cell.block)
        (sameRow + sameColumn + sameBlock)
            .filterIsInstance<SolvedCell>()
            .mapTo(EnumSet.noneOf(SudokuNumber::class.java)) { it.value }
            .let { visibleValues -> enumIntersect(cell.candidates, visibleValues) }
            .takeUnless { it.isEmpty() }
            ?.let { toRemove -> RemoveCandidates(cell, toRemove) }
    }